src/NotificationEventMailing.php

```diff
@@ -131,22 +131,14 @@ public static function send(array $data)
                     }
                 }
 
-                // Add custom header for mail grouping in reader
-                $mmail->AddCustomHeader(
-                    str_replace(
-                        [
-                            '%uuid',
-                            '%itemtype',
-                            '%items_id'
-                        ],
-                        [
-                            Config::getUuid('notification'),
-                            $current->fields['itemtype'],
-                            $current->fields['items_id']
-                        ],
-                        "In-Reply-To: <GLPI-%uuid-%itemtype-%items_id>"
-                    )
-                );
+                // Add custom header (In-Reply-To) for mail grouping in reader
+                // Add custom header (References)  for mail grouping in reader (for Gmail)
+                // see https://workspaceupdates.googleblog.com/2019/03/threading-changes-in-gmail-conversation-view.html
+                $current_item = new $current->fields['itemtype']();
+                $current_item->getFromDB($current->fields['items_id']);
+                $email_ref = $current_item->getItemMessageId($current->fields['event']);
```

Comment:
There are some edge cases where `itemtype` is not valid (was not valid on creation or is not valid as related plugin is not active anymore), or where `items_id` refers to an object that does not exists anymore (a notification related to a deleted object for instance).

1. The concat/replace logic should probably be located in current class.
2. It should handle the case where itemtype is not valid (or null).
3. Loading the item is not necessary, and I guess there are sometimes notifications that are sent using `$items_id = 0`.
4. There is already a method used to generate MessageId header, but here the generated string is for a `Reference`, so method name should probably contains `Reference` instead of `Id` to prevent misunderstandings.

In `NotificationEventMailing`:
```php
    public static function getItemMessageReference(?string $itemtype, int $items_id, ?string $event): string
    {
        $ref_event = 'none';
        if (!empty($event)) {
            $ref_event = !empty($itemtype) && is_a($itemtype, CommonDBTM::class, true)
                ? $itemtype::getMessageReferenceEvent($event)
                : $event;
        }

        return sprintf(
            'GLPI_%s-%s-%s/%s@%s',
            Config::getUuid('notification'),
            $itemtype,
            $items_id,
            $ref_event,
            php_uname('n')
        );
    }
```

Reply:
agree with that